[Solax X1] Rework request and respond processing

SteWers opened the pull request:
## Description:

Complete rework of the request cycle and the respond processing. This is more reliable and reusable for other and further requests.
Right now the serial number of the converter is additionally requested and displayed in the WebUI.

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [ ] The code change is tested and works with Tasmota core ESP32 V.2.0.2.1
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

SteWers committed:
[Solax X1] Rework request and respond processing

Complete rework of the request cycle and the respond processing. This is more reliable and reusable for more and further requests.
Right now the serial number of the converter is requested and displayed in the WebUI.

arendst commented on tasmota/xnrg_12_solaxX1.ino:
```diff
@@ -131,15 +107,16 @@ uint8_t functionCode[1] = {0x00};
+uint8_t solaxX1_send_retry = 20;
+uint8_t solaxX1_queryData_count = 0;
+uint8_t solaxX1_QueryID_count = 240;
+uint8_t solaxX1SerialNumber[16] = {0x6e, 0x2f, 0x61}; // "n/a"
```
Pls move all these global variables into the ``solaxX1`` struct to fix future global variable naming issues.

SteWers replied:
@arendst Thanks for your advice. As I am working on extensions to request more information from the inverter I will implement your requested changes in my current work. So you can merge this PR and I will start a new PR in the next few days. Otherwise I can publish a new commit to this PR with all the changes and additions.

SteWers replied:
@arendst The work is done and I commited the changes.

SteWers committed:
[Solax X1] rework 02/2022 final

- rework of the request cycle
- more reliable and reusable respond processing for more and further requests
- periodically request of IDinfo data to get the converters serial number and display it on the WebUI
- restructure the usage and names of variables; especially the global ones, to prevent naming conflicts
- on demand request of IDinfo data via command `EnergyConfig12 ReadIDinfo`
- on demand request of config data via command `EnergyConfig12 ReadConfig`
- many other code optimisations

arendst commented on tasmota/xnrg_12_solaxX1.ino:
```diff
@@ -492,21 +579,26 @@ bool Xnrg12(uint8_t function)
+      solaxX1_DrvInit();
+      break;
+    case FUNC_COMMAND:
+      if (XNRG_12 == XdrvMailbox.index) { // "EnergyConfig12 <command>"
```
No need to check for XNRG_12 as these functions are only called if XNRG_12 is the active energy service driver. There can be only ONE active energy service driver.

arendst commented on tasmota/xnrg_12_solaxX1.ino:
```diff
@@ -104,295 +101,364 @@ struct SOLAXX1 {
+        AddLog(LOG_LEVEL_INFO, PSTR("SX1: WFrqProtectRestrictive: %d"), (DataRead[71] << 8) | DataRead[72]);
+        AddLog(LOG_LEVEL_INFO, PSTR("SX1: WQuDelayTimer: %d sec"), (DataRead[73] << 8) | DataRead[74]);
+        AddLog(LOG_LEVEL_INFO, PSTR("SX1: WFreqActivePowerDelayTimer: %d ms"), (DataRead[75] << 8) | DataRead[76]);
+
```
My my my am I glad this driver is an optional driver. It wastes way too much code space for non-relevant information.

You might want to dive into the use of ``%*_f`` instead of dtostrfd().

SteWers committed:
[Solax X1] rework 02/2022 final 2

- removed command index
- new syntax: on demand request of IDinfo data via command `EnergyConfig ReadIDinfo`
- new syntax: on demand request of config data via command `EnergyConfig ReadConfig`
- replaced `dtostrfd()` with `%*_f`

arendst commented:
A perfect example how not to write a driver.
